python/tests/test_ld_matrix.py

```diff
@@ -0,0 +1,559 @@
+
+            # compute stat
+            stats[a_idx, b_idx] = summary_func(w_AB, w_Ab, w_aB, n)
+
+            if print_weights:
+                print(a_idx, b_idx, int(w_AB), int(w_Ab), int(w_aB), int(n), sep="\t")
+            # create weights matrix
+            if norm_method is NormMethod.HAP_WEIGHTED:
+                hap_freq = hap_counts / n
+                weights[a_idx, b_idx] = hap_freq[a_idx, b_idx]
+            elif norm_method is NormMethod.AF_WEIGHTED:
+                p_a = hap_counts.sum(1) / n
+                p_b = hap_counts.sum(0) / n
+                weights[a_idx, b_idx] = p_a[a_idx] * p_b[b_idx]
+            elif norm_method is NormMethod.TOTAL:
+                weights[a_idx, b_idx] = 1 / (
+                    (n_a - (1 if polarized else 0)) * (n_b - (1 if polarized else 0))
+                )
+
+    return stats, weights
+
+
+@dataclass
+class SiteMatrixIndex:
+    """
+    This object stores index values for the output result matrix and
+    for the sites array (stored on this object).
+
+    For each index array, we store information for sites that are shared
+    between rows and columns, and sites that differ between rows and columns.
+    """
+
+    n_rows: int  # number of row sites
+    n_cols: int  # number of column sites
+    # Union of row/column sites (in order)
+    sites: List[int] = field(default_factory=list)
+    # Result matrix indices for shared/diff between rows/columns
+    rshr_matrix: List[int] = field(default_factory=list)
+    cshr_matrix: List[int] = field(default_factory=list)
+    rdiff_matrix: List[int] = field(default_factory=list)
+    cdiff_matrix: List[int] = field(default_factory=list)
+    # Index into the sites array on this object for rows/columns/shared
+    # This index also maps to the data gathered by get_state. (Important!!)
+    shr_sites: List[int] = field(default_factory=list)
+    rdiff_sites: List[int] = field(default_factory=list)
+    cdiff_sites: List[int] = field(default_factory=list)
+
+
+def check_sites(sites, max_sites):
+    """Validate the specified site ids.
+
+    We require that sites are:
+
+    1) Within the boundaries of available sites in the tree sequence
+    2) Sorted
+    3) Non-repeating
+
+    Raises an exception if any error is found.
+
+    :param sites: 1d array of sites to validate
+    :param max_sites: Number of sites in the tree sequence, the upper
+                      bound value for site ids.
+    """
+    if sites is None or len(sites) == 0:
+        raise ValueError("No sites provided")
+    i = 0
+    for i in range(len(sites) - 1):
+        if sites[i] < 0 or sites[i] >= max_sites:
+            raise ValueError(f"Site out of bounds: {sites[i]}")
+        if sites[i] >= sites[i + 1]:
+            raise ValueError(f"Sites not sorted: {sites[i], sites[i + 1]}")
+    if sites[-1] < 0 or sites[-1] >= max_sites:
+        raise ValueError(f"Site out of bounds: {sites[i + 1]}")
+
+
+def get_matrix_indices(row_sites, col_sites):
```

Comment:
I think this would be more naturally done as the constructor to SiteMatrixIndex, so that we call `` idx = SiteMatrixIndex(row_sites, col_sites)`` in client code. Here, I found myself going up and down the file to try and figure out what the fields were actually doing, where we should really just encapsulate everything in the SiteMatrixIndex class.